src/ray/raylet/local_object_manager.cc

```diff
@@ -298,14 +298,27 @@ void LocalObjectManager::AddSpilledUrls(
 }
 
 void LocalObjectManager::AsyncRestoreSpilledObject(
-    const ObjectID &object_id, const std::string &object_url,
+    const ObjectID &object_id, const NodeID &node_id,
     std::function<void(const ray::Status &)> callback) {
-  RAY_LOG(DEBUG) << "Restoring spilled object " << object_id << " from URL "
-                 << object_url;
   if (objects_pending_restore_.count(object_id) > 0) {
     // If the same object is restoring, we dedup here.
     return;
   }
+
+  if (!spilled_objects_url_.contains(object_id)) {
```

Comment:
Shouldn't we check whether node_id is non-nil instead? This seems like it will break S3 spilling.

Reply:
I think this is the main difference from our thought. So you are suggesting to just follow the previous protocol when we use non-file-based spilling? (cuz I thought we will just follow this protocol for all external storages).